Add toggles for individual Phantoon patterns (#62)

NobodyNada opened the pull request:
As specified in #62, allow the user to select a subset of Phantoon patterns to practice.

Implemented using hooks in Phantoon's AI that override the default RNG-based pattern chooser with a custom implementation that chooses patterns from the user-selected subset. If all the patterns are enabled (or all are disabled), the custom logic is disabled and Phantoon's regular RNG code is run instead -- so this feature should not impact
Phantoon's behavior or pattern probabilities when not in use.

Also includes an option to skip the long intro, because I got tired of sitting through it. With the "Skip Intro" option enabled, Phantoon begins the fight (starting from his wavy-fade-in animation) immediately when you enter the boss room.

idlechild reviewed (commented):
A lot of work here, I'll need to try this out. I like the idea of fast phantoon entry, and it makes sense you might want to practice some random patterns without practicing all of them.

I'll need to take a closer look to ensure RNG was preserved by default.  It looks like you tried to do that.

NobodyNada commented:
I definitely tried to preserve RNG; if the user has not modified Phantoon's behavior it should be identical to the vanilla game unless I missed something silly. The RNG hooks simply execute a verbatim copy of the patched-out code (see the `.no_manip` branches) unless the user has selected a custom set of patterns, with no additional RNG calls or anything like that.

Also, even if the user has overridden Phantoon's behavior we should still preserve the number of RNG calls per frame (we always patch out one RNG call but make up for it with one call at the beginning of `choose_phantoon_pattern`).

InsaneFirebat commented:
I like the intro skip here. That's another win for players that rely on presets as their savestates. I would have liked to have seen the 1st/2nd round settings kept, but this is a pretty good trade-off.

Unfortunately, this doesn't preserve the odds for rare patterns unless all patterns are enabled. Both the mid-left and fast-right patterns of the second round are appearing at a significant rate when any pattern is turned off. I'm fighting Phantoon with 3-4 missiles per round and getting one or two rare patterns per fight, with just one pattern disabled. With both slow patterns disabled I'm seeing quite a lot of them.

idlechild commented:
Once you start toggling off certain patterns, I'm not sure it is practical to maintain the vanilla odds.  The user might not want the vanilla odds in that situation anyway.

With an entire submenu just for phantoon, perhaps it's possible to also have the 1st/2nd round options (maybe not, I haven't given it much thought so it might be super complicated).

Another possibility is we add a toggle to control vanilla rng vs more balanced rng, and then any time you toggle vanilla rng on it automatically toggles all patterns on.  This would make it more clear that turning off patterns changes the rng.  Also you could then toggle vanilla rng off and leave all patterns on if desired.

NobodyNada commented:
My reasoning was: the rare patterns are rare enough that the user could just toggle them off they didn’t want to practice them. I can find a way to restore the vanilla odds if needed, though it would make the pattern chooser quite a bit more complicated so I wasn’t sure if that was necessary or even desired.

Separate 1st/2nd round options isn’t complicated at all. The reason I didn’t implement it is mainly UX: it would have meant twice as many menu options to scroll through. But if that feature is valuable to people’s workflows, I can bring it back.

InsaneFirebat commented:
Having the 1st/2nd round options would let you disable the rare patterns. Without it, you'd be disabling two extra patterns for the first round just to get them out of the second.

Can we move that intro skip flag to SRAM and copy the menu pointer (`phan_skip_intro`) into the Game > Cutscenes menu?

NobodyNada commented:
That makes sense, I didn't think about that! Thanks, I'll make those changes when I get the time (it might be a few days as I've got a busy rest-of-the-week).

idlechild commented:
There's no rush :)

NobodyNada commented:
I've implemented those changes:
<img width="521" alt="Screen Shot 2022-02-06 at 12 46 49 AM" src="https://user-images.githubusercontent.com/19418817/152673717-dc81f733-cacc-47ea-9b1c-5c820bc5b812.png">

The toggles are now separated by round, and the intro skip is moved to the cutscenes menu. As before, if the patterns for a round are not modified, we use vanilla RNG to preserve vanilla odds. I've also updated the documentation (but not the changelog).

idlechild commented:
Looks good.

I'll likely pull this into my candidate near the end of the week in preparation of 2.4.1 release, although it might not happen until the following week.  I've got some GT classic practice to do :)
